### DISCUSSION TITLE
Communicate between hook and endpoint extensions

### DISCUSSION THREAD

**Discussion by rhyst** [directus/directus#8582] [Archive]
Is it possible to communicate between a hook and an endpoint extension? Basically my endpoint does some caching and I want it to invalidate the cache when the items change. A hooks seems like the way to go but is there a way to pass information between them? Is there was a way to register a hook from an endpoint extension?

Thanks!

**Comment by nickrum** [comment]
You could implement your cache in a separate file and import it from your hook and your endpoint. This is of course not ideal, but I don't think there is a built-in way to communicate between extensions yet. (cc @rijkvanzanten)

**Reply by rhyst** [reply]
Thanks for the reply. I've ended up just writing some stuff to `global` as that seemed easier than managing more files. 

Is there a better way to write an extension that requires hooks and endpoints? For my purposes something like `registerExtension({ registerHook, registerEndpoint }) { ... }` would be ideal but it seems like they are deliberately separated at the moment.

**Reply by nickrum** [reply]
Good idea, that works as well of course.

Yes, you are right, they are deliberately separated to enforce _Separation of Concerns_. You could implement your own custom endpoint inside a hook by using one of the "init" hooks as they provide access to the express app, but wouldn't recommend doing that. We should come up with a clean way to support this totally valid use case.
